### ISSUE TITLE
Media content caching in a swarm

### ISSUE THREAD

**Issue by wolfie1910** [tootsuite/mastodon#6511]
Hello... I have an issue when I run multiple web services in swarm. When running several web services, each web service has it is own media content cache, and they can't access each other's caches.
How can I solve the problem that the client would be able to display media content and avatars when connecting to any web service?

* * * *

- [ ] I searched or browsed the repo’s other issues to ensure this is not a duplicate.
- [ ] This bug happens on a [tagged release](https://github.com/tootsuite/mastodon/releases) and not on `master` (If you're a user, don't worry about this).

**Comment by Gargron** [comment]
With swarm I'm assuming the different containers are on different physical machines? Because on a single machine you'd have all of them access the same volume, as is default in docker-compose.yml. Your options are to use Minio (self-hosted), Amazon S3 or Wasabi.

**Comment by wolfie1910** [comment]
yea, sure on different machines. I am also experiencing some bugs, such as when accessing the instance the web is frozen (just showing blue empty page) or there are alot of bugs in rendering the web page....

**Comment by wolfie1910** [comment]
@Gargron can u help how to assign the volumes using Minio...? and maybe answer the question just above (rendering problem tho no errors in logs)

**Comment by Gargron** [comment]
> such as when accessing the instance the web is frozen

You need to run `assets:precompile` on every machine, or copy over public/packs/manifest.json to every machine after precompile on one of them. That's one thing, and that's about static assets only (CSS/JS)

>can u help how to assign the volumes using Minio...?

For user uploads, that's where Minio comes in. You're no longer using a volume when you use Minio. You set it up like Amazon S3 and your containers send data to Minio instead of writing files to a local directory.

**Comment by wolfie1910** [comment]
@Gargron So, I should remove the volume for the /public/system?
Also, I have the problem in .env.production with  S3_ENDPOINT field. When I check docker logs for my Minio container, I get two IPs, but I have tried different combinations to enter them in production file, and still web logs give me  "URI::InvalidURIError (bad URI(is not URI?)" error....

**Comment by andrwmorph** [comment]
If you use S3 you can ditch the /public/system endpoint.
If you don't want to use minio, you could compile the assets into the docker image or you could run a global service to individually build them in the same location on each host.

**Comment by wolfie1910** [comment]
@andrwmorph @Gargron  yes, I know all of that... My point is now what S3_ENDPOINT value shall I put...?
I am using Minio as a service on a server, not in a container....

p.s. If I dont enter any value, the web sees the configuration as AW3 Amazon.

**Comment by andrwmorph** [comment]
@wolfie1910  This is what I have for us-east-1:
`S3_REGION=us-east-1
S3_PROTOCOL=https
S3_HOSTNAME=s3.amazonaws.com`

from [AWS docs](https://docs.aws.amazon.com/general/latest/gr/rande.html#s3_region)

**Comment by nightpool** [comment]
The endpoint should just be the URL that's the host of your minio server.
